### PR TITLE
Octavia - List Amphoare

### PR DESCRIPTION
**Pull request by lingxiankong**
For #1179

octavia doc: <https://developer.openstack.org/api-ref/load-balancer/v2/index.html#list-amphora>
octavia source code: <https://github.com/openstack/octavia/blob/b0210e4312655423d8a58931e74d5e7b2ed4974c/octavia/api/v2/controllers/amphora.py#L57>

### CODE REVIEW

**Comment by kiwik** [comment]
recheck

**Comment by lingxiankong** [comment]
recheck

**Comment by lingxiankong** [comment]
@jtopjian hi, this PR is ready for review :-)

File: openstack/loadbalancer/v2/amphorae/results.go [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+	LoadbalancerID string `json:"loadbalancer_id"`
+
+	// The management IP of the amphora.
+	LBNetworkIP string `json:"lb_network_ip"`
+
+	// The ID of the amphora resource in the compute system.
+	ComputeID string `json:"compute_id"`
+
+	// The IP address of the Virtual IP (VIP).
+	VIP string `json:"ha_ip"`
```
Comment (jtopjian) [inline openstack/loadbalancer/v2/amphorae/results.go]: `VIP` => `HAIP`

File: openstack/loadbalancer/v2/amphorae/results.go [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+	// The ID of the Virtual IP (VIP) port.
+	VIPPortID string `json:"ha_port_id"`
```
Comment (jtopjian) [inline openstack/loadbalancer/v2/amphorae/results.go]: `VIPPortID` => `HAPortID`

File: openstack/loadbalancer/v2/amphorae/results.go [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+	// The date the certificate for the amphora expires.
+	CertExpiration string `json:"cert_expiration"`
+
+	// Whether the certificate is in the process of being replaced.
+	IsCertBusy bool `json:"cert_busy"`
```
Comment (jtopjian) [inline openstack/loadbalancer/v2/amphorae/results.go]: `IsCertBusy` => `CertBusy`

File: openstack/loadbalancer/v2/amphorae/results.go [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+	// The role of the amphora. One of STANDALONE, MASTER, BACKUP.
+	Role string `json:"role"`
+
+	// The status of the amphora. One of: BOOTING, ALLOCATED, READY, PENDING_CREATE, PENDING_DELETE, DELETED, ERROR.
+	Status string `json:"status"`
+
+	// The vrrp port’s ID in the networking system.
+	VrrpPortID string `json:"vrrp_port_id"`
```
Comment (jtopjian) [inline openstack/loadbalancer/v2/amphorae/results.go]: `VrrpPortID` => `VRRPPortID`

File: openstack/loadbalancer/v2/amphorae/results.go [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+	// The address of the vrrp port on the amphora.
+	VrrpIP string `json:"vrrp_ip"`
```
Comment (jtopjian) [inline openstack/loadbalancer/v2/amphorae/results.go]: `VrrpIP` => `VRRPIP`

File: openstack/loadbalancer/v2/amphorae/results.go [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+	// The bound interface name of the vrrp port on the amphora.
+	VrrpInterface string `json:"vrrp_interface"`
```
Comment (jtopjian) [inline openstack/loadbalancer/v2/amphorae/results.go]: `VrrpInterface` => `VRRPInterface`

File: openstack/loadbalancer/v2/amphorae/results.go [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+
+	// The vrrp group’s ID for the amphora.
+	VrrpID int `json:"vrrp_id"`
```
Comment (jtopjian) [inline openstack/loadbalancer/v2/amphorae/results.go]: `VrrpID` => `VRRPID`

File: openstack/loadbalancer/v2/amphorae/results.go [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+
+	// The priority of the amphora in the vrrp group.
+	VrrpPriority int `json:"vrrp_priority"`
```
Comment (jtopjian) [inline openstack/loadbalancer/v2/amphorae/results.go]: `VrrpPriority` => `VRRPPriority`

File: openstack/loadbalancer/v2/amphorae/requests.go [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+// the Amphorae attributes you want to see returned. SortKey allows you to
+// sort by a particular attribute. SortDir sets the direction, and is
+// either `asc' or `desc'. Marker and Limit are used for pagination.
+type ListOpts struct {
+	LoadbalancerID string `q:"loadbalancer_id"`
+	ImageID        string `q:"image_id"`
+	Role           string `q:"role"`
+	Status         string `q:"status"`
+	VIPPortID      string `q:"ha_port_id"`
+	VrrpPortID     string `q:"vrrp_port_id"`
```
Comment (jtopjian) [inline openstack/loadbalancer/v2/amphorae/requests.go]: `VrrpPortID` => `VRRPPortID`

File: openstack/loadbalancer/v2/amphorae/requests.go [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+// the API. Filtering is achieved by passing in struct field values that map to
+// the Amphorae attributes you want to see returned. SortKey allows you to
+// sort by a particular attribute. SortDir sets the direction, and is
+// either `asc' or `desc'. Marker and Limit are used for pagination.
+type ListOpts struct {
+	LoadbalancerID string `q:"loadbalancer_id"`
+	ImageID        string `q:"image_id"`
+	Role           string `q:"role"`
+	Status         string `q:"status"`
+	VIPPortID      string `q:"ha_port_id"`
```
Comment (jtopjian) [inline openstack/loadbalancer/v2/amphorae/requests.go]: `VIPPortID` => `HAPortID`

File: openstack/loadbalancer/v2/amphorae/results.go [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+	return gophercloud.ExtractNextURL(s.Links)
+}
+
+// IsEmpty checks whether a AmphoraPage struct is empty.
+func (r AmphoraPage) IsEmpty() (bool, error) {
+	is, err := ExtractAmphoare(r)
+	return len(is) == 0, err
+}
+
+// ExtractAmphoare accepts a Page struct, specifically a AmphoraPage
```
Comment (jtopjian) [inline openstack/loadbalancer/v2/amphorae/results.go]: `ExtractAmphoare` => `ExtractAmphorae`

File: openstack/loadbalancer/v2/amphorae/results.go [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+// struct, and extracts the elements into a slice of Amphora structs. In
+// other words, a generic collection is mapped into a relevant slice.
+func ExtractAmphoare(r pagination.Page) ([]Amphora, error) {
```
Comment (jtopjian) [inline openstack/loadbalancer/v2/amphorae/results.go]: `ExtractAmphoare` => `ExtractAmphorae`

File: openstack/loadbalancer/v2/amphorae/results.go [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+	err := r.ExtractInto(&s)
+	if err != nil {
+		return "", err
+	}
+	return gophercloud.ExtractNextURL(s.Links)
+}
+
+// IsEmpty checks whether a AmphoraPage struct is empty.
+func (r AmphoraPage) IsEmpty() (bool, error) {
+	is, err := ExtractAmphoare(r)
```
Comment (jtopjian) [inline openstack/loadbalancer/v2/amphorae/results.go]: ExtractAmphoare => ExtractAmphorae

File: openstack/loadbalancer/v2/amphorae/doc.go [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+	listOpts := amphorae.ListOpts{
+		LoadbalancerID: "6bd55cd3-802e-447e-a518-1e74e23bb106",
+	}
+
+	allPages, err := amphorae.List(octaviaClient, listOpts).AllPages()
+	if err != nil {
+		panic(err)
+	}
+
+	allAmphorae, err := amphorae.ExtractAmphoare(allPages)
```
Comment (jtopjian) [inline openstack/loadbalancer/v2/amphorae/doc.go]: ExtractAmphoare => ExtractAmphorae

File: openstack/loadbalancer/v2/amphorae/testing/requests_test.go [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+func TestListAmphorae(t *testing.T) {
+	th.SetupHTTP()
+	defer th.TeardownHTTP()
+	HandleAmphoraListSuccessfully(t)
+
+	pages := 0
+	err := amphorae.List(fake.ServiceClient(), amphorae.ListOpts{}).EachPage(func(page pagination.Page) (bool, error) {
+		pages++
+
+		actual, err := amphorae.ExtractAmphoare(page)
```
Comment (jtopjian) [inline openstack/loadbalancer/v2/amphorae/testing/requests_test.go]: ExtractAmphoare => ExtractAmphorae

File: openstack/loadbalancer/v2/amphorae/testing/requests_test.go [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+}
+
+func TestListAllAmphorae(t *testing.T) {
+	th.SetupHTTP()
+	defer th.TeardownHTTP()
+	HandleAmphoraListSuccessfully(t)
+
+	allPages, err := amphorae.List(fake.ServiceClient(), amphorae.ListOpts{}).AllPages()
+	th.AssertNoErr(t, err)
+	actual, err := amphorae.ExtractAmphoare(allPages)
```
Comment (jtopjian) [inline openstack/loadbalancer/v2/amphorae/testing/requests_test.go]: ExtractAmphoare => ExtractAmphorae

**Comment by lingxiankong** [comment]
@jtopjian, field names modified, thanks for review. You are correct about 'amphorae' :-)

File: openstack/loadbalancer/v2/amphorae/doc.go (hunk 14 above)
Comment (jtopjian) [reply openstack/loadbalancer/v2/amphorae/doc.go]: @lingxiankong One last one :)

File: openstack/loadbalancer/v2/amphorae/results.go [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+	ComputeID string `json:"compute_id"`
+
+	// The IP address of the Virtual IP (VIP).
+	HAIP string `json:"ha_ip"`
+
+	// The ID of the Virtual IP (VIP) port.
+	HAPortID string `json:"ha_port_id"`
+
+	// The date the certificate for the amphora expires.
+	CertExpiration string `json:"cert_expiration"`
```
Comment (jtopjian) [inline openstack/loadbalancer/v2/amphorae/results.go]: This should be a time field.

File: openstack/loadbalancer/v2/amphorae/testing/fixtures.go [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+            "created_at": "2018-08-09T23:44:31",
+            "ha_ip": "10.0.0.6",
+            "ha_port_id": "35254b63-9361-4561-9b8f-2bb4e3be60e3",
+            "id": "45f40289-0551-483a-b089-47214bc2a8a4",
+            "image_id": "5d1aed06-2624-43f5-a413-9212263c3d53",
+            "lb_network_ip": "192.168.0.6",
+            "loadbalancer_id": "882f2a9d-9d53-4bd0-b0e9-08e9d0de11f9",
+            "role": "MASTER",
+            "status": "READY",
+            "updated_at": "2018-08-09T23:51:06",
```
Comment (jtopjian) [inline openstack/loadbalancer/v2/amphorae/testing/fixtures.go]: Looks like this field is missing from the `Amphorae` result.

File: openstack/loadbalancer/v2/amphorae/testing/requests_test.go [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+func TestListAllAmphorae(t *testing.T) {
+	th.SetupHTTP()
+	defer th.TeardownHTTP()
+	HandleAmphoraListSuccessfully(t)
+
+	allPages, err := amphorae.List(fake.ServiceClient(), amphorae.ListOpts{}).AllPages()
+	th.AssertNoErr(t, err)
+	actual, err := amphorae.ExtractAmphorae(allPages)
+	th.AssertNoErr(t, err)
+	th.AssertEquals(t, 2, len(actual))
```
Comment (jtopjian) [inline openstack/loadbalancer/v2/amphorae/testing/requests_test.go]: We'll want to make sure the time fields are unmarshalled properly here.

**Comment by lingxiankong** [comment]
@jtopjian, thanks for your review, I will make the change according to you suggestion.

**Pull request merged**